addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java

```diff
@@ -0,0 +1,783 @@
+
+            try {
+                while (!listenerStopped & ((nhcMessage = nhcIn.readLine()) != null)) {
+                    readMessage(nhcMessage);
+                }
+            } catch (IOException e) {
+                if (!listenerStopped) {
+                    // this is not an communication stop triggered from outside this runnable
+                    logger.warn("Niko Home Control: IO error in listener");
+                    // the IO has stopped working, so we need to close cleanly and try to restart
+                    restartCommunication();
+                }
+            }
+        }
+
+    };
+
+    /**
+     * Method that interprets all feedback from Niko Home Control and calls appropriate handling methods.
+     *
+     * @param nhcMessage message read from Niko Home Control.
+     */
+    private void readMessage(String nhcMessage) {
+
+        Gson gson = new Gson();
+
+        logger.debug("Niko Home Control: received json {}", nhcMessage);
+
+        try {
+            // first try with simple reply form, a hashmap in the data field
+            NHCSimpleCmdAck nhcSimpleCmdAck = gson.fromJson(nhcMessage, NHCSimpleCmdAck.class);
+            if ("systeminfo".equals(nhcSimpleCmdAck.cmd)) {
+                systemInfo(nhcSimpleCmdAck);
+            } else if ("startevents".equals(nhcSimpleCmdAck.cmd)) {
+                startEvents(nhcSimpleCmdAck);
+            } else if ("executeactions".equals(nhcSimpleCmdAck.cmd)
+                    && (Integer.valueOf(nhcSimpleCmdAck.data.get("error")).equals(0))) {
+                logger.debug("Niko Home Control: execute action success");
+            } else {
+                logger.debug("Niko Home Control: not acted on json {}", nhcMessage);
+            }
+        } catch (JsonSyntaxException je) {
+            // if that JSON cannot be converted, it should be an array of hashmaps in the data field
+            NHCCmdAck nhcCmdAck = gson.fromJson(nhcMessage, NHCCmdAck.class);
+            if ("listlocations".equals(nhcCmdAck.cmd)) {
+                listLocations(nhcCmdAck);
+            } else if ("listactions".equals(nhcCmdAck.cmd)) {
+                listActions(nhcCmdAck);
+            } else if ("listactions".equals(nhcCmdAck.event)) {
+                for (HashMap<String, String> action : nhcCmdAck.data) {
+                    int id = Integer.valueOf(action.get("id"));
+                    setActionState(id, Integer.valueOf(action.get("value1")));
+                }
+            } else {
+                logger.debug("Niko Home Control: not acted on json {}", nhcMessage);
+            }
+        }
+    }
+
+    /**
+     * Cleanup socket when the communication with Niko Home Control IP-interface is closed.
+     *
+     */
+    public synchronized void stopCommunication() {
+        listenerStopped = true;
+
+        if (nhcSocket != null) {
+            try {
+                nhcSocket.close();
+            } catch (IOException e) {
+            }
```

Comment:
Most of the times its better to explicitly ignore an exception, for this location it makes sense. You could name the exception `ignore` and add a comment why you ignored id.